ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -629,6 +646,165 @@ private IndexValue findKey(StoreKey key, FileSpan fileSpan, EnumSet<IndexEntryTy
+                + fileSpan.getEndOffset());
+        segmentsMapToSearch = indexSegments.subMap(indexSegments.floorKey(fileSpan.getStartOffset()), true,
+            indexSegments.floorKey(fileSpan.getEndOffset()), true).descendingMap();
+        metrics.segmentSizeForExists.update(segmentsMapToSearch.size());
+      }
+      int segmentsSearched = 0;
+      for (Map.Entry<Offset, IndexSegment> entry : segmentsMapToSearch.entrySet()) {
+        segmentsSearched++;
+        logger.trace("Index : {} searching index with start offset {}", dataDir, entry.getKey());
+        NavigableSet<IndexValue> values = entry.getValue().find(key);
+        if (values != null) {
+          if (result == null) {
+            result = new LinkedList<>();
+          }
+          Iterator<IndexValue> it = values.descendingIterator();
+          while (it.hasNext()) {
+            IndexValue value = it.next();
+            if ((types.contains(IndexEntryType.DELETE) && value.isDelete()) || (types.contains(IndexEntryType.UNDELETE)
+                && value.isUndelete()) || (types.contains(IndexEntryType.TTL_UPDATE) && !value.isDelete()
+                && value.isTTLUpdate()) || (types.contains(IndexEntryType.PUT) && value.isPut())) {
+              // Add a copy of the value to the result since we return a modifiable list to the caller.
+              IndexValue newValue =
+                  new IndexValue(value.getSize(), value.getOffset(), value.getFlags(), value.getExpiresAtMs(),
+                      value.getOperationTimeInMs(), value.getAccountId(), value.getContainerId(),
+                      value.getLifeVersion());
+              result.add(newValue);
+            }
+          }
+        }
+      }
+      metrics.segmentsAccessedPerBlobCount.update(segmentsSearched);
+    } finally {
+      context.stop();
+    }
+    if (result != null) {
+      logger.trace("Index: {} Returninng values {}", dataDir, result);
+    }
+    return result;
+  }
+
+  /**
+   * Ensure that the previous {@link IndexValue}s is structured correctly for undeleting the {@code key}.
+   * <p/>
+   * Undelete should be permitted only when the first record is a Put and last record is a Delete, and the Put record
+   * hasn't expired yet.
+   * @param key the key to be undeleted.
+   * @param values the previous {@link IndexValue}s in reversed order.
+   * @param lifeVersion lifeVersion for the undelete record, it's only valid when in recovery or replication.
+   */
+  void validateSanityForUndelete(StoreKey key, List<IndexValue> values, short lifeVersion) throws StoreException {
+    if (values == null || values.isEmpty()) {
+      throw new StoreException("Id " + key + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
+    }
+    if (!IndexValue.hasLifeVersion(lifeVersion)) {
+      validateSanityForUndeleteWithoutLifeVersion(key, values);
+      return;
+    }
+    IndexValue firstValue = values.get(values.size() - 1);
+    IndexValue lastValue = values.get(0);
```

Comment:
I get you point here but I would suggest reverse values before hand which makes logic easier to understand and more consistent with your java docs of this method.

Reply:
I think the order in list is correct, i will change the variable name.